src/vehicle-class-validator/vehicle.ts

```diff
@@ -0,0 +1,73 @@
+import { InternalServerErrorException } from '@nestjs/common';
+import { IntersectionType } from '@nestjs/mapped-types';
+import { IsIn, IsNumber, IsString, ValidateIf } from 'class-validator';
+import { VehicleEnum, VehicleType } from '../common/vehicle';
+import { VehicleEntity } from '../vehicle/vehicle.entity';
+
+export class VehicleDto {
+  @IsIn(VehicleEnum.values)
+  type!: VehicleType;
+
+  @IsString()
+  plate!: string;
+
+  @IsString()
+  modelName!: string;
+}
+
+export class CarDto extends VehicleDto {
+  @ValidateIf((o) => o.type === VehicleEnum.car.value)
+  @IsNumber()
+  seats!: number;
+}
+
+export class MotorbikeDto extends VehicleDto {
+  @ValidateIf((o) => o.type === VehicleEnum.moto.value)
+  @IsNumber()
+  engineCapacity!: number;
+}
+
+export class CreateVehicleDto extends IntersectionType(CarDto, MotorbikeDto) {}
+
+type encoderVehicle<T extends VehicleDto> = (v: T) => VehicleEntity;
+
+const encodeVehicleToEntity: encoderVehicle<VehicleDto> = (
+  v: VehicleDto
+): VehicleEntity => {
+  const ve = new VehicleEntity();
+  ve.modelName = v.modelName;
+  ve.plate = v.plate;
+  return ve;
+};
+
+const encodeCarToEntity: encoderVehicle<CarDto> = (
+  v: CarDto
+): VehicleEntity => {
+  const ve = encodeVehicleToEntity(v);
+  ve.type = VehicleEnum.car.value;
+  ve.seats = v.seats;
+  return ve;
+};
+
+const encodeMotorbikeToEntity = (v: MotorbikeDto): VehicleEntity => {
+  const ve = encodeVehicleToEntity(v);
+  ve.type = VehicleEnum.moto.value;
+  ve.engineCapacity = v.engineCapacity;
+  return ve;
+};
+
+const ENCODER: Record<VehicleType, encoderVehicle<any>> = {
+  [VehicleEnum.car.value]: encodeCarToEntity,
+  [VehicleEnum.moto.value]: encodeMotorbikeToEntity
+};
+
+export const encodeToEntityFactory = (encoderResolver = ENCODER) => {
```

Comment:
Qui ti fa pagare il fatto di non avere il DTO come tagged union.
Abbiamo provato a riscriverlo per evitare di lanciare l'eccezione in fondo:

```typescript
export const encodeToEntityFactory = () => {
  return (v: VehicleDto) => {
    switch(v.type) {
      case VehicleEnum.car.value:
        return encodeCarToEntity(v as CarDto);
      case VehicleEnum.moto.value:
        return encodeMotorbikeToEntity(v as MotorbikeDto)
    }
  };
};
```

Purtroppo con una struttura dati di questo tipo, i `case` non riescono a raffinare il tipo specifico del DTO e sono costretto ad utilizzare `as` (che bypassa il type system).

Reply:
esatto, motivo per cui dicevo che la modellazione dei tipi mi piace molto con io-ts